Scripts/Process/MomentProcessor.cs

```diff
@@ -0,0 +1,156 @@
+﻿namespace VRTK.Core.Process
+{
+    using UnityEngine;
+
+    /// <summary>
+    /// The MomentProcessor loops through the given moments and executes their Process method on the given Unity game loop moment.
+    /// </summary>
+    public class MomentProcessor : MonoBehaviour
+    {
+        /// <summary>
+        /// The MomentToProcess is the point in the Unity game loop when to execute the processes.
+        /// </summary>
+        public enum MomentToProcess
+        {
+            /// <summary>
+            /// Never execute the processes.
+            /// </summary>
+            None,
+            /// <summary>
+            /// Execute the processes in the FixedUpdate physics part of the Unity game loop.
+            /// </summary>
+            OnFixedUpdate,
+            /// <summary>
+            /// Executes the processes in the Update game logic part of the Unity game loop.
+            /// </summary>
+            OnUpdate,
+            /// <summary>
+            /// Executes the processes in the LateUpdate game logic part of the Unity game loop.
+            /// </summary>
+            OnLateUpdate,
+            /// <summary>
+            /// Executes the processes in the camera PreCull scene rendering part of the Unity game loop.
+            /// </summary>
+            OnPreCull,
+            /// <summary>
+            /// Executes the processes in the camera PreRender scene rendering part of the Unity game loop.
+            /// </summary>
+            OnPreRender
+        }
+
+        [Tooltip("The moment in which to process the given processes.")]
+        public MomentToProcess momentToProcess = MomentToProcess.OnPreRender;
+        [Tooltip("A collection of Moments to process.")]
+        public MomentProcess[] processes = new MomentProcess[0];
+
+        protected MomentToProcess subscribedMoment;
+
+        protected virtual void OnEnable()
+        {
+            subscribedMoment = MomentToProcess.None;
+            ManageSubscriptions();
+        }
+
+        protected virtual void OnDisable()
+        {
+            UnsubscribeMoment();
+        }
+
+        protected virtual void FixedUpdate()
+        {
+            if (momentToProcess == MomentToProcess.OnFixedUpdate)
+            {
+                Process();
+            }
+        }
+
+        protected virtual void Update()
+        {
+            if (momentToProcess == MomentToProcess.OnUpdate)
+            {
+                Process();
+            }
+        }
+
+        protected virtual void LateUpdate()
+        {
+            if (momentToProcess == MomentToProcess.OnLateUpdate)
+            {
+                Process();
+            }
+            ManageSubscriptions();

```

Comment:
I wonder if this needs to be done in multiple places to pick up a change in the `momentToProcess` earlier.

We should probably discuss a generic solution to support variables that show up in the inspector to change at edit-time but that allow changing at runtime, too. I think so far we are doing this "check regularly" approach as well as an "provide setter methods" approach - need to settle on one or even on a new solution.

Reply:
I don't think we should have setter methods for anything that is a public inspector parameter because you can't change them in the editor at runtime, which means the only thing you can do for things like this is have a polling check.

The reason it only happens in the LateUpdate at the moment is it's not the end of the world if you have to wait for the next frame when changing the param

Reply:
[This](http://adammechtley.com/2014/06/property-backing-field-drawer/) + [this](https://docs.microsoft.com/en-us/dotnet/api/system.runtime.compilerservices.callermembernameattribute?view=netframework-4.7.2).